### PR TITLE
stardust-xr-server: init at 0.44.1

### LINKED ISSUE #299584: Package request: stardustxr-server
**Project description**

Stardust XR (informally known as Stardust) is an XR display server for Linux-based distros (possibly most unix-based too, but untested). Unlike display servers of the past that were run on bounded 2D screens, Stardust has to break convention in order to account for unbounded 3D space in any potential virtual or real environment.

**Metadata**

* homepage URL: https://stardustxr.org/
* source URL: https://github.com/StardustXR/server
* license: gpl2+
* platforms: linux

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Pandapip1**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/299584

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/st/stardust-xr-server/package.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    # This is not ideal, the original approach was to fetch the exact cmake
+    # file version that was wanted from GitHub directly, but at least this way it comes from Nixpkgs.. so meh
+    cp ${cpm-cmake}/share/cpm/CPM.cmake $sk/build/cpm/CPM_0.32.2.cmake
+  '';
+
+  passthru.updateScript = writeScript "envision-update" ''
+    source ${builtins.head (gitUpdater { })}
+    cp $tmpdir/Cargo.lock ./pkgs/by-name/en/envision-unwrapped/Cargo.lock
+  '';
```
Comment (Pandapip1) [inline pkgs/by-name/st/stardust-xr-server/package.nix:67]: Todo: fix the envision reference I forgot to change

File: pkgs/by-name/st/stardust-xr-server/package.nix [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  buildInputs = [
+    openxr-loader
+    libGL
+    mesa
+    xorg.libX11
+    fontconfig
+    libxkbcommon
+  ];
+
+  LIBCLANG_PATH = "${libclang.lib}/lib";
```
Comment (Aleksanaa) [inline pkgs/by-name/st/stardust-xr-server/package.nix:55]: Try `rustPlatform.bindgenHook` (add to `nativeBuildInputs`), if that works remove all things with `clang`.

File: pkgs/by-name/st/stardust-xr-server/package.nix [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      update-source-version \
+          "$UPDATE_NIX_ATTR_PATH" \
+          "$new_version" \
+          --ignore-same-hash
+      cp "$tmpdir/Cargo.lock" ./pkgs/by-name/st/stardust-xr-server/Cargo.lock
+    '';
+  });
+
+  meta = {
+    description = "A wayland compositor and display server for 3D applications";
```
Comment (Aleksanaa) [inline pkgs/by-name/st/stardust-xr-server/package.nix:94]:
```suggestion
    description = "Wayland compositor and display server for 3D applications";
```

File: pkgs/by-name/st/stardust-xr-server/package.nix [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    hash = "sha256-sCatpWDdy7NFWOWUARjN3fZMDVviX2iV79G0HTxfYZU=";
+  };
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "smithay-0.3.0" = "sha256-dxAgTGW+xxnL+vA6j2Ng02F1zt/Y5VaSxP9xg8jfMy8=";
+      "stardust-xr-0.14.1" = "sha256-fmRb46s0Ec8wnoerBh4JCv1WKz2of1YW+YGwy0Gr/yQ=";
+    };
+  };
```
Comment (Aleksanaa) [inline pkgs/by-name/st/stardust-xr-server/package.nix:40]:
```suggestion
  };

```

File: pkgs/by-name/st/stardust-xr-server/package.nix [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+    llvmPackages.libcxxClang
+  ];
```
Comment (Aleksanaa) [inline pkgs/by-name/st/stardust-xr-server/package.nix:45]:
```suggestion
  ];

```

Comment (Pandapip1) [reply pkgs/by-name/st/stardust-xr-server/package.nix:42]: Not according to `nixfmt-rfc-style`.

File: pkgs/by-name/st/stardust-xr-server/package.nix (hunk 4 above)
Comment (Pandapip1) [reply pkgs/by-name/st/stardust-xr-server/package.nix:36]: Not according to `nixfmt-rfc-style`.

File: pkgs/by-name/st/stardust-xr-server/package.nix (hunk 2 above)
Comment (Pandapip1) [reply pkgs/by-name/st/stardust-xr-server/package.nix]: That works. Thanks!

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]
Build fails
```
  CMake Warning at build/cpm/CPM_0.32.2.cmake:76 (message):
    CPM: Your project is using an unstable development version of CPM.cmake.
    Please update to a recent release if possible.  See
    https://github.com/cpm-cmake/CPM.cmake for details.
  Call Stack (most recent call first):
    cmake/CPM.cmake:21 (include)
    CMakeLists.txt:111 (include)


  CMake Warning:
    Manually-specified variables were not used by the project:

      CMAKE_ASM_COMPILER
      CMAKE_ASM_FLAGS


  In file included from /build/cargo-vendor-dir/stereokit-sys-2.5.6/StereoKit/StereoKitC/libraries/sk_gpu.cpp:2:
  /build/cargo-vendor-dir/stereokit-sys-2.5.6/StereoKit/StereoKitC/libraries/sk_gpu.h: In function 'bool skg_read_file(const char*, void**, size_t*)':
  /build/cargo-vendor-dir/stereokit-sys-2.5.6/StereoKit/StereoKitC/libraries/sk_gpu.h:4817:15: warning: ignoring return value of 'size_t fread(void*, size_t, size_t, FILE*)' declared with attribute 'warn_unused_result' [-Wunused-result]
   4817 |         fread (*out_data, *out_size, 1, fp);
        |         ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /build/cargo-vendor-dir/stereokit-sys-2.5.6/StereoKit/StereoKitC/platforms/linux.cpp:10:10: fatal error: X11/extensions/Xfixes.h: No such file or directory
     10 | #include <X11/extensions/Xfixes.h>
        |          ^~~~~~~~~~~~~~~~~~~~~~~~~
  compilation terminated.
  make[2]: *** [CMakeFiles/StereoKitC.dir/build.make:860: CMakeFiles/StereoKitC.dir/StereoKitC/platforms/linux.cpp.o] Error 1
  make[2]: *** Waiting for unfinished jobs....
  make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/StereoKitC.dir/all] Error 2
  make: *** [Makefile:136: all] Error 2
  thread 'main' panicked at /build/cargo-vendor-dir/cmake-0.1.50/src/lib.rs:1098:5:

  command did not execute successfully, got: exit status: 2

  build script failed, must exit now
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by Pandapip1** [comment]
Welp, time to figure out why that is.

**Comment by Pandapip1** [comment]
Fixed

**Review by FliegendeWurst — APPROVED** [review]

**Review by Aleksanaa — APPROVED** [review]

**Comment by Aleksanaa** [comment]
Please verify it still works.

**Comment by Pandapip1** [comment]
I have re-tested and can confirm it still works.

**Pull request merged**